midware/middleware.go

```diff
@@ -0,0 +1,155 @@
+package midware
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+
+	"github.com/google/uuid"
+)
+
+// DefaultTraceHeader the default header when TraceHeaders is given an empty
+// string instead of a valid header name.
+var DefaultTraceHeader = "X-Request-Id"
+
+// PathOverrides is middleware which overrides handling for a specified set of
+// http request paths.  Each entry in a PathOverrides map is an http request
+// path and the associated handler will be used to serve that path instead of
+// allowing the middleware's "natural" inner handler to serve the request.
+//
+// PathOverrides does not support overriding subtrees (paths ending with '/')
+// in the way that http.ServeMux supports path patterns.  Keys in PathOverrides
+// are expected to be complete, rooted paths.
+type PathOverrides map[string]http.Handler
+
+// Wrap implements the Middleware interface.
+func (m PathOverrides) Wrap(next http.Handler) http.Handler {
+	return &pathOverridesHandler{m, next}
+}
+
+type pathOverridesHandler struct {
+	m    PathOverrides
+	next http.Handler
+}
+
+func (h *pathOverridesHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if route, ok := h.m[r.URL.Path]; ok {
+		route.ServeHTTP(w, r)
+		return
+	}
+	h.next.ServeHTTP(w, r)
+}
+
+// ServerResponseHeader returns a middleware that renders the given sequence of
+// server components (presumably in "software[/version]" format) and includes
+// them in the Server response header.  Any secondary components which are
+// supplied in addition to primary will be rendered in sequence and delimited
+// by a single whitespace.  Any component which produces an empty string will
+// be ignored.  ServerResponseHeader overwrites any Server header that was set
+// earlier (by another middleware).
+//
+// ServerResponseHeader renders the primary component once immediately when
+// called and will panic if it returns an empty string.  Later at runtime the
+// behavior of a ServerResponseHeader middleware is not defined when the result
+// of rendering the sequence of components is an empty string.  It is
+// recommended that the primary component have a knowably non-empty rendered
+// value (e.g. pass the result of ServerFixed called with const arguments).
```

Comment:
The semantics described in the paragraph are maybe a bit complex. I could change `primary` to just be a string that is required to be non-empty. But that makes `ServerFixed` less useful, or I would need to include an alternate version of `ServerFixed` that just returns a string and while retain the current version that returns `func() string`.

Reply:
I guess I am tempted to introduce a second `ServerFixed` style function. I guess these are the signatures I imagine,

```go
func ServerFixed(name string, version string) string
func ServerFixedFunc(name string, version string) func() string
```

Reply:
I did end changing the signature of this function and I also clarified and fixed the semantics a bit.